Include/cpython/critical_section.h

```diff
@@ -0,0 +1,134 @@
+#ifndef Py_CPYTHON_CRITICAL_SECTION_H
+#  error "this header file must not be included directly"
+#endif
+
+// Python critical sections
```

Comment:
Maybe move this doc to Doc/c-api/init.rst, or merge it with Doc/c-api/init.rst?

Reply:
I think the file documentation is still useful even with the public `Doc/c-api/init.rst` docs. The file docs talk more about specific internals like `PyThreadState.critical_section`.